### PR TITLE
JSDK-3022 DTX for Opus codec.

### PR DESCRIPTION
**Pull request by manjeshbhargav**
@makarandp0 @ceaglest 

This pull request implements the ConnectOptions API spec changes specified [here](https://code.hq.twilio.com/client/video-sdk-api/pull/73). The LocalTrackPublishOptions API spec changes will be implemented in a separate release.

## Checklist

- [x] Implementation
- [x] Manual verification with React App
- [x] Unit tests
- [x] Integration tests
- [x] Changelog entry
- [x] +1 from at least one reviewer

**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,14 +2,29 @@ The Twilio Programmable Video SDKs use [Semantic Versioning](http://www.semver.o
...
 **Support for 1.x will cease on December 4th, 2020**. This branch will only receive fixes for critical issues until that date. Check [this guide](https://www.twilio.com/docs/video/migrating-1x-2x) when planning your migration to 2.x. For details on the 1.x changes, go [here](https://github.com/twilio/twilio-video.js/blob/support-1.x/CHANGELOG.md).
 
-2.7.3 (in progress)
+2.8.0 (in progress)
 ===================
 
+New Features
+------------
+
+- Enabled discontinuous transmission (DTX) in the Opus audio codec, which will result in
```
Comment (idelgado) [inline CHANGELOG.md:11]: > in the Opus audio codec by default,

**Review by idelgado — CHANGES_REQUESTED** [review]
Should I expect to see tests here too?

For example,

When I join with DTX enabled to a Group Room I see my sender bitrate lower.
When I join P2P and both peers enables DTX
When I join P2P and 1 peer enables DTX

**Comment by manjeshbhargav** [comment]
@idelgado 

> Should I expect to see tests here too?
>
> For example,
>
> When I join with DTX enabled to a Group Room I see my sender bitrate lower.
When I join P2P and both peers enables DTX
When I join P2P and 1 peer enables DTX

We do have tests to verify the application of `usedtx=1` to the local sdps. However, I don't know if there is a way to write deterministic integration tests because you have to synthesize speech and silence to test the codec behavior. Also, the bitrate changes take some time, so I need to use timeouts to calculate the bitrate, which can be flaky.

**Comment by idelgado** [comment]
@manjeshbhargav what if you were to use disabled? Does the bitrate look different when DTX is on for the sender?

I think integration tests make sense here. Maybe this is something QE should add but it should be somewhere and the behavior should be validated and codified by tests.

**Comment by manjeshbhargav** [comment]
@idelgado 

Yes, the bitrate does go down by a lot during silence, but not all the way to 0. We would have to come up with a bitrate threshold estimate to compare against during silence.

**Comment by manjeshbhargav** [comment]
@idelgado I tried testing with a synthetic audio track, and opus seems to behave differently than microphone tracks. The bandwidth reduction is very less in Chrome and Safari and non-existent in Firefox. But, if I use a microphone track, dtx operates as expected. I’m not sure how to simulate microphone tracks in integration tests. This feature most likely will require manual verification.

**Comment by manjeshbhargav** [comment]
@idelgado I tried testing with a synthetic audio track, and opus seems to behave differently than microphone tracks. The bandwidth reduction is very less in Chrome and Safari and non-existent in Firefox. But, if I use a microphone track, dtx operates as expected. I’m not sure how to simulate microphone tracks in integration tests. This feature most likely will require manual verification.

**Review by idelgado — APPROVED** [review]
Please have QE validate the scenarios I listed.

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,14 +2,29 @@ The Twilio Programmable Video SDKs use [Semantic Versioning](http://www.semver.o
...
-2.7.3 (in progress)
+2.8.0 (in progress)
 ===================
 
+New Features
+------------
+
+- Enabled discontinuous transmission (DTX) in the Opus audio codec by default, which
+  will result in bandwidth and CPU savings during silence and background noise. You
+  can control this feature using the ConnectOptions property `preferredAudioCodecs`. (JSDK-3022)
```
Comment (makarandp0) [inline CHANGELOG.md:13]: I think we should add more details about what should application consider disabling dtx.

File: CHANGELOG.md [hunk 3]
```diff
@@ -2,14 +2,29 @@ The Twilio Programmable Video SDKs use [Semantic Versioning](http://www.semver.o
 
 **Support for 1.x will cease on December 4th, 2020**. This branch will only receive fixes for critical issues until that date. Check [this guide](https://www.twilio.com/docs/video/migrating-1x-2x) when planning your migration to 2.x. For details on the 1.x changes, go [here](https://github.com/twilio/twilio-video.js/blob/support-1.x/CHANGELOG.md).
 
-2.7.3 (in progress)
+2.8.0 (in progress)
```
[the remark this answers is not in the document]
Comment (manjeshbhargav) [reply CHANGELOG.md:5]: @makarandp0 Yes. We need to do that.

File: CHANGELOG.md (hunk 2 above)
Comment (manjeshbhargav) [reply CHANGELOG.md:13]: @makarandp0 I want to test with a non-speech audio track (like music). I'll update the changelog after that.

File: CHANGELOG.md [hunk 4]
```diff
@@ -2,14 +2,29 @@ The Twilio Programmable Video SDKs use [Semantic Versioning](http://www.semver.o
...
 ===================
 
+New Features
+------------
+
+- Enabled discontinuous transmission (DTX) in the Opus audio codec, which will result in
+  bandwidth and CPU savings during silence and background noise. You can control this
```
[the remark this answers is not in the document]
Comment (manjeshbhargav) [reply CHANGELOG.md]: @makarandp0 I don't think that's needed.

**Review by ceaglest — COMMENTED** [review]
Overall the changes look good.

From an API compatibility standpoint, do you think it is too aggressive to enable DTX by default in a minor version? I am not totally opposed since it is an optimization, and the primary use case is voice traffic.

It would be nice to observe the reduction in bandwidth after enabling DTX in a test. Could you inject silence and then patch in a voice recording of the say verb?

**Comment by manjeshbhargav** [comment]
@ceaglest ,

I don't think it is too aggressive to enable DTX by default, since its an optimization. Also, making this opt-in requires customer education and proactive steps on our part to get them to adopt it.

**Review by ceaglest — APPROVED** [review]
Ack, we should be prepared for the possibility that some customers will not want DTX and will reach out to us about it after trying 2.8.0. This PR is a good start on the education front.

Use your discretion on the SDK level test coverage.

**Pull request merged**